### DISCUSSION TITLE
Feel++ support cmake presets

### DISCUSSION THREAD

**Discussion by prudhomm** [feelpp/feelpp#1705] [Announcements]
## Introduction

Feel++ now supports CMake Presets [[1]](https://cmake.org/cmake/help/v3.21/manual/cmake-presets.7.html) and vscode supports it as well [[2]](https://github.com/microsoft/vscode-cmake-tools/blob/develop/docs/cmake-presets.md).

We have now a new system of configuration which allows to set up configurations and builds of feelpp.
In vscode replaces the cmake-variants by the cmake-tools but it is necessary to have cmake 3.21. 
This version of cmake is installed on  ubutu 20.04 via snap. 
```
sudo snap install --classic cmake
# then update .bashrc or .zshrc
export PATH=/snap/cmake/current/bin:$PATH
```

## Feel++ components

Feel++ now enjoys three more components aside from `feelpp`, `toolboxes` and `mor`, namely the python wrapper of the exiting components:
* `pyfeelpp` the python wrapper for `feelpp`
* `pyfeelpp-toolboxes` the python wrapper for `toolboxes`
* `pyfeelpp-mor` the python wrapper for `mor`

This allows to build separately the python wrappers as long as the corresponding components and their dependencies are installed.

## CmakePresets Configurations

The available Presets configurations are defined by the components currently, but they will be extended very soon:
We have:

* `default` on configuration which compiles and installs everything
* `feelpp` compiles and installs only the Feel++ library
* `pyfeelpp` compiles and installs only the pyfeelpp component
* `toolboxes` compiles and installs only the Feel++ toolboxes 
* `pyfeelpp-toolboxes` compiles and installs only the Feel++ toolboxes python wrapper 
* `mor` compiles and installs only the Feel++ MOR 
* `pyfeelpp-mor` compiles and installs only the Feel++ MOR python wrapper 

to execute a preset, just type for example
```
cd <feelpp top source dir>
mkdir build
cmake --preset feelpp 
```
and it will automatically configure the `feelpp` preset and build it in `build/feelpp`

**Comment by romainhild** [comment]
@prudhomm with `cmake --preset testsuite` the build folder is not used and every build folder in testsuite is created in the top source dir. Is that normal ?

**Reply by prudhomm** [reply]
I will check asap

**Reply by prudhomm** [reply]
it does not seem to be the case:

```json
 {
            "name": "testsuite",
            "inherits": "default",
            "displayName": "testsuite",
            "description": "Build only the Feel++ testsuite Component",
            "binaryDir": "${sourceDir}/build/testsuite",
            "cacheVariables": {
                "FEELPP_COMPONENT": "testsuite"
            }
        },
```

**Reply by prudhomm** [reply]
@romainhild could you check again ?

**Reply by romainhild** [reply]
ok indeed, I used an old version of cmake, so I guess the preset was not really taken into account.

**Comment by romainhild** [comment]
If one wish to use a preset but want to have the build files somewhere else, one can use `cmake --preset feelpp -B <path/to/build>` for example.
